[GTK] Create WebKitWebExtension

ItsJamie9494 opened the pull request:
#### 68f61154cacb9920c6c79f3d0e4de23d1f6d3270
<pre>
[GTK] Create WebKitWebExtension
<a href="https://webkit.org/b/285379">https://webkit.org/b/285379</a>

Reviewed by Adrian Perez de Castro and Patrick Griffis.

This creates a GLib API of WebExtension, along with some utility functions needed for Linux-specific pathways. This new API is only built when ENABLE_2022_GLIB_API is enabled,
preventing a conflict with the older WebKitWebExtension API that was renamed.

As well, this fixes WEBKIT_WEB_EXTENSION_MATCH_PATTERN_ERROR as it originally pointed to webkit_web_extension_error_quark(). Both WEBKIT_WEB_EXTENSION_MATCH_PATTERN_ERROR and WEBKIT_WEB_EXTENSION_ERROR now are properly defined with their respective quarks.

Tests: Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebExtension.cpp
       Tools/TestWebKitAPI/glib/CMakeLists.txt
       Tools/TestWebKitAPI/glib/WebKitGLib/WebExtensionUtilities.cpp
       Tools/TestWebKitAPI/glib/WebKitGLib/WebExtensionUtilities.h
* Source/WebCore/platform/graphics/Icon.h:
* Source/WebCore/platform/graphics/gtk/IconGtk.cpp:
(WebCore::Icon::Icon):
(WebCore::Icon::create):
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/Shared/API/APIData.h:
(API::Data::createWithoutCopying):
* Source/WebKit/Shared/Extensions/gtk/WebExtensionUtilitiesGtk.cpp: Copied from Source/WebKit/UIProcess/API/glib/WebKitPrivate.h.
(WebKit::availableScreenScales):
* Source/WebKit/Shared/Extensions/wpe/WebExtensionUtilitiesWPE.cpp: Copied from Source/WebKit/UIProcess/API/glib/WebKitPrivate.h.
(WebKit::availableScreenScales):
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/glib/WebKitError.cpp:
(webkit_web_extension_error_quark):
* Source/WebKit/UIProcess/API/glib/WebKitError.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitPrivate.cpp:
(toWebKitWebExtensionError):
* Source/WebKit/UIProcess/API/glib/WebKitPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp: Added.
(webkitWebExtensionGetProperty):
(webkit_web_extension_class_init):
(webkitWebExtensionCreate):
(webkit_web_extension_new):
(webkit_web_extension_get_manifest_version):
(webkit_web_extension_supports_manifest_version):
(webkit_web_extension_get_default_locale):
(webkit_web_extension_get_display_name):
(webkit_web_extension_get_display_short_name):
(webkit_web_extension_get_display_version):
(webkit_web_extension_get_display_description):
(webkit_web_extension_get_display_action_label):
(webkit_web_extension_get_icon):
(webkit_web_extension_get_action_icon):
(webkit_web_extension_get_version):
(webkit_web_extension_get_requested_permissions):
(webkit_web_extension_get_optional_permissions):
(webkit_web_extension_get_requested_permission_match_patterns):
(webkit_web_extension_get_optional_permission_match_patterns):
(webkit_web_extension_get_all_requested_match_patterns):
(webkit_web_extension_get_has_background_content):
(webkit_web_extension_get_has_service_worker_background_content):
(webkit_web_extension_get_has_modular_background_content):
(webkit_web_extension_get_has_persistent_background_content):
(webkit_web_extension_get_has_injected_content):
(webkit_web_extension_get_has_options_page):
(webkit_web_extension_get_has_override_new_tab_page):
(webkit_web_extension_get_has_commands):
(webkit_web_extension_get_has_content_modification_rules):
* Source/WebKit/UIProcess/API/glib/WebKitWebExtension.h.in: Added.
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionInternal.h: Copied from Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPatternPrivate.h.
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPatternPrivate.h:
* Source/WebKit/UIProcess/API/glib/webkit.h.in:
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
(WebKit::WebExtension::WebExtension):
* Source/WebKit/UIProcess/Extensions/gtk/WebExtensionGtk.cpp: Added.
(WebKit::WebExtension::WebExtension):
(WebKit::WebExtension::resourceDataForPath):
(WebKit::WebExtension::recordError):
(WebKit::WebExtension::iconForPath):
(WebKit::WebExtension::bestIcon):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebExtension.cpp: Added.
(createGBytes):
(testExtensionCreationFromDirectory):
(testDisplayStringParsing):
(testDefaultLocaleParsing):
(testDisplayStringParsingWithLocalization):
(testActionParsing):
(testContentScriptsParsing):
(testPermissionsParsing):
(testBackgroundParsing):
(testBackgroundPreferredEnvironmentParsing):
(testOptionsPageParsing):
(testURLOverridesParsing):
(testContentSecurityPolicyParsing):
(testWebAccessibleResourcesV2):
(testWebAccessibleResourcesV3):
(testCommandsParsing):
(testDeclarativeNetRequestParsing):
(testExternallyConnectableParsing):
(beforeAll):
(afterAll):
* Tools/TestWebKitAPI/glib/CMakeLists.txt:
* Tools/TestWebKitAPI/glib/WebKitGLib/WebExtensionUtilities.cpp: Copied from Source/WebKit/UIProcess/API/glib/WebKitPrivate.h.
(TestWebKitAPI::Util::makePNGData):
* Tools/TestWebKitAPI/glib/WebKitGLib/WebExtensionUtilities.h: Copied from Source/WebKit/UIProcess/API/glib/WebKitPrivate.h.

Canonical link: <a href="https://commits.webkit.org/301078@main">https://commits.webkit.org/301078@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2f1b090102a201e4271d53d3fdd4b17e62908efa

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124499 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44177 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34911 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131337 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76479 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126376 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44894 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52758 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94707 "") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62813 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127453 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35783 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111344 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75285 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34721 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29500 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74816 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105535 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29731 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134002 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51370 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39195 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103186 "") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51777 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107566 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102971 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26298 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48332 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26594 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/48313 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51240 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57030 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50651 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54010 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52331 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitError.h.in:
```diff
@@ -227,6 +262,9 @@ webkit_javascript_error_quark                  (void);
 WEBKIT_API GQuark
 webkit_snapshot_error_quark                    (void);
 
+WEBKIT_API GQuark
+webkit_web_extension_error_quark               (void);
```
```suggestion
#if ENABLE(2022_GLIB_API)
WEBKIT_API GQuark
webkit_web_extension_error_quark               (void);
#endif
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
```diff
@@ -0,0 +1,1160 @@
+ * 
+ * Since: 2.52
+ */
+GList* webkit_web_extension_get_optional_permission_match_patterns(WebKitWebExtension* extension)
```
Can we just return normal arrays instead of `GLists`? Emmanuele [suggests](https://www.bassi.io/articles/2023/02/20/bindable-api-2023/)  to just never use `GList` in public APIs.

ItsJamie9494 replied:
I think it would almost make more sense to have a MatchPatternList instead of a normal C array? As far as I can tell, WebKit doesn't really use C arrays for anything but string lists, everything else is either a custom List holder or a GList

mcatanzaro replied:
A null-terminated array seems pretty safe to me. Why add a new type if you don't need to?

mcatanzaro commented on Source/WebKit/PlatformGTK.cmake:
```diff
@@ -212,6 +212,15 @@ if (ENABLE_2022_GLIB_API)
     )
 endif ()
 
+if (ENABLE_2022_GLIB_API)
+    list(APPEND WebKitGTK_HEADER_TEMPLATES
+        ${WEBKIT_DIR}/UIProcess/API/glib/WebKitWebExtension.h.in
+    )
+    list(APPEND WebKit_SOURCES
+        ${WEBKIT_DIR}/UIProcess/API/glib/WebKitWebExtension.cpp
+    )
+endif ()
```
I see you were careful to not put GTK-specific stuff into this file. What more is needed to get it building for WPE? Would be nice to at least add it to the build system so that it gets compiled, even if it's not ready yet. Otherwise we'll have to split out all that documentation and maintain it separately for each port in order to get separate Since tags, which is not worth it.

ItsJamie9494 replied:
Internally, there's some code that relies on Gdk (for icons). This code is, unless I am forgetting a use, either creating a pixbuf to store the icons, or is using Gdk to get screen scaling. While these aren't exposed publicly, WPE specific versions of these functions would likely need to be written.

ItsJamie9494 replied:
Did some testing, we can compile WPE with the WebExtension headers just fine, but as described above, the WK_WEB_EXTENSIONS flag shouldn't be enabled on WPE as that'll break. So on WPE, it'll just be left as the stub APIs that don't do anything

TingPing replied:
That sounds fine for now, we can add icon support to WPEPlatform later. There is `wpe_screen_get_scale()` for scaling.

ItsJamie9494 replied:
I'll add support for the scaling function on WPE then

TingPing commented on Source/WebKit/UIProcess/API/glib/WebKitError.h.in:
```diff
@@ -36,7 +36,10 @@ G_BEGIN_DECLS
 #define WEBKIT_SNAPSHOT_ERROR                    webkit_snapshot_error_quark ()
 #define WEBKIT_USER_CONTENT_FILTER_ERROR         webkit_user_content_filter_error_quark ()
-#define WEBKIT_WEB_EXTENSION_MATCH_PATTERN_ERROR webkit_web_extension_error_quark ()
+#if ENABLE(2022_GLIB_API)
+#define WEBKIT_WEB_EXTENSION_ERROR               webkit_web_extension_error_quark ()
+#endif
+#define WEBKIT_WEB_EXTENSION_MATCH_PATTERN_ERROR webkit_web_extension_match_pattern_error_quark ()
```
I read this as `webkit_web_extension_error_quark()` wasn't previously defined, so `WEBKIT_WEB_EXTENSION_MATCH_PATTERN_ERROR` has been broken?

Maybe worth documenting in the commit at least.

ItsJamie9494 replied:
Yeah, it was broken, the quark function was properly named later in this file and just the #define here was wrong. Will add a note to the commit about that.

TingPing commented on Source/WebKit/UIProcess/API/glib/WebKitError.cpp:
```diff
@@ -112,12 +112,30 @@ GQuark webkit_snapshot_error_quark()
  *
  * Returns: web extension match pattern error domain.
+ * 
+ * Since: 2.48
```
`2.52`

ItsJamie9494 replied:
I'll set this to 2.52, but the actual domain here has existed since 2.48. If anything it's a bug that this wasn't added in originally.

mcatanzaro replied:
No, set it to 2.48. Patrick surely just didn't notice.

TingPing commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtensionInternal.h:
```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (C) 2024 Igalia S.L.
```
`2025`

TingPing commented on Tools/TestWebKitAPI/glib/WebKitGLib/WebExtensionUtilities.h:
```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (C) 2024 Igalia S.L.
```
`2025`

TingPing commented on Source/WebKit/UIProcess/API/glib/WebKitError.h.in:
```diff
@@ -166,6 +199,8 @@ typedef enum {
  *
  * Enum values used to denote errors happening when creating a #WebKitWebExtensionMatchPattern
+ * 
+ * Since: 2.48
```
`2.52`

aperezdc commented on Source/WebKit/UIProcess/Extensions/gtk/WebExtensionGtk.cpp:
```diff
@@ -0,0 +1,239 @@
+            auto urlEncodedString = path.substring(commaPosition);
+            auto decodedString = URL(urlEncodedString).string();
+            return API::Data::create(decodedString.utf8().span());
+        }
```
This could probably be replaced with `WebCore::DataURLDecoder::decode()` 🤔

ItsJamie9494 replied:
Likely, I wasn't aware of that class. Now, the Cocoa implementation doesn't use DataURLDecoder either, and my main goal is for the implementation details to try and match Cocoa as much as possible. But, maybe it could work for both as well.

xeenon replied:
I also wasn't aware of the `DataURLDecoder` class. I support adopting it for WebExtension.

ItsJamie9494 replied:
I'll open a PR for WebExtensionCocoa soon

aperezdc commented on Source/WebKit/UIProcess/Extensions/gtk/WebExtensionGtk.cpp:
```diff
@@ -0,0 +1,239 @@
+        return makeUnexpected(nullptr);
+    }
+
+    auto rawData = FileSystem::readEntireFile((resourceURL).fileSystemPath());
```
Unneeded parenthesis:

```suggestion
    auto rawData = FileSystem::readEntireFile(resourceURL.fileSystemPath());
```

Or am I missing something?

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
```diff
@@ -0,0 +1,1203 @@
+
+    WebKitWebExtension* object = WEBKIT_WEB_EXTENSION(g_object_new(WEBKIT_TYPE_WEB_EXTENSION, nullptr));
+    object->priv->extension = WTFMove(extension);
+    return object;
```
This prevents being able to create instances with `g_object_new(WEBKIT_TYPE_WEB_EXTENSION, ..., nullptr)` outside of internal WebKit code. Consumers of the API will expect being able to do that, and also being able to use the generic instantiation mechanism is needed for language bindings through GObject-Introspection.

I would suggest:

- Making the `WebKitWebExtension:path` a construct-only property.
- Implementing the `GInitable` interface to be able to report errors during object initialization.
- Moving the code that creates the `WebKit::WebExtension` into the `GInitable.init` virtual method.
- Making `webkit_web_extension_new()` call `g_initable_new()` instead under the hood.

ItsJamie9494 replied:
Yes, this makes sense. Will do, thank you

aperezdc reviewed (changes requested):
This is looking very good, and I think the API additions are good to go. I have a few suggestions about implementation details, please take a look.

aperezdc reviewed (approved):
Some code to test this in the MiniBrowser would be great, rather than in Epiphany, but I don't want to block landing this on that. I think this can be landed, and adding some bits and bobs to the MiniBrowser can be done later on.

ItsJamie9494 committed:
[GTK] Create WebKitWebExtension
https://webkit.org/b/285379

Reviewed by Adrian Perez de Castro and Patrick Griffis.

This creates a GLib API of WebExtension, along with some utility functions needed for Linux-specific pathways. This new API is only built when ENABLE_2022_GLIB_API is enabled,
preventing a conflict with the older WebKitWebExtension API that was renamed.

As well, this fixes WEBKIT_WEB_EXTENSION_MATCH_PATTERN_ERROR as it originally pointed to webkit_web_extension_error_quark(). Both WEBKIT_WEB_EXTENSION_MATCH_PATTERN_ERROR and WEBKIT_WEB_EXTENSION_ERROR now are properly defined with their respective quarks.

Tests: Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebExtension.cpp
       Tools/TestWebKitAPI/glib/CMakeLists.txt
       Tools/TestWebKitAPI/glib/WebKitGLib/WebExtensionUtilities.cpp
       Tools/TestWebKitAPI/glib/WebKitGLib/WebExtensionUtilities.h
* Source/WebCore/platform/graphics/Icon.h:
* Source/WebCore/platform/graphics/gtk/IconGtk.cpp:
(WebCore::Icon::Icon):
(WebCore::Icon::create):
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/Shared/API/APIData.h:
(API::Data::createWithoutCopying):
* Source/WebKit/Shared/Extensions/gtk/WebExtensionUtilitiesGtk.cpp: Copied from Source/WebKit/UIProcess/API/glib/WebKitPrivate.h.
(WebKit::availableScreenScales):
* Source/WebKit/Shared/Extensions/wpe/WebExtensionUtilitiesWPE.cpp: Copied from Source/WebKit/UIProcess/API/glib/WebKitPrivate.h.
(WebKit::availableScreenScales):
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/glib/WebKitError.cpp:
(webkit_web_extension_error_quark):
* Source/WebKit/UIProcess/API/glib/WebKitError.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitPrivate.cpp:
(toWebKitWebExtensionError):
* Source/WebKit/UIProcess/API/glib/WebKitPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp: Added.
(webkitWebExtensionGetProperty):
(webkit_web_extension_class_init):
(webkitWebExtensionCreate):
(webkit_web_extension_new):
(webkit_web_extension_get_manifest_version):
(webkit_web_extension_supports_manifest_version):
(webkit_web_extension_get_default_locale):
(webkit_web_extension_get_display_name):
(webkit_web_extension_get_display_short_name):
(webkit_web_extension_get_display_version):
(webkit_web_extension_get_display_description):
(webkit_web_extension_get_display_action_label):
(webkit_web_extension_get_icon):
(webkit_web_extension_get_action_icon):
(webkit_web_extension_get_version):
(webkit_web_extension_get_requested_permissions):
(webkit_web_extension_get_optional_permissions):
(webkit_web_extension_get_requested_permission_match_patterns):
(webkit_web_extension_get_optional_permission_match_patterns):
(webkit_web_extension_get_all_requested_match_patterns):
(webkit_web_extension_get_has_background_content):
(webkit_web_extension_get_has_service_worker_background_content):
(webkit_web_extension_get_has_modular_background_content):
(webkit_web_extension_get_has_persistent_background_content):
(webkit_web_extension_get_has_injected_content):
(webkit_web_extension_get_has_options_page):
(webkit_web_extension_get_has_override_new_tab_page):
(webkit_web_extension_get_has_commands):
(webkit_web_extension_get_has_content_modification_rules):
* Source/WebKit/UIProcess/API/glib/WebKitWebExtension.h.in: Added.
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionInternal.h: Copied from Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPatternPrivate.h.
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPatternPrivate.h:
* Source/WebKit/UIProcess/API/glib/webkit.h.in:
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
(WebKit::WebExtension::WebExtension):
* Source/WebKit/UIProcess/Extensions/gtk/WebExtensionGtk.cpp: Added.
(WebKit::WebExtension::WebExtension):
(WebKit::WebExtension::resourceDataForPath):
(WebKit::WebExtension::recordError):
(WebKit::WebExtension::iconForPath):
(WebKit::WebExtension::bestIcon):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebExtension.cpp: Added.
(createGBytes):
(testExtensionCreationFromDirectory):
(testDisplayStringParsing):
(testDefaultLocaleParsing):
(testDisplayStringParsingWithLocalization):
(testActionParsing):
(testContentScriptsParsing):
(testPermissionsParsing):
(testBackgroundParsing):
(testBackgroundPreferredEnvironmentParsing):
(testOptionsPageParsing):
(testURLOverridesParsing):
(testContentSecurityPolicyParsing):
(testWebAccessibleResourcesV2):
(testWebAccessibleResourcesV3):
(testCommandsParsing):
(testDeclarativeNetRequestParsing):
(testExternallyConnectableParsing):
(beforeAll):
(afterAll):
* Tools/TestWebKitAPI/glib/CMakeLists.txt:
* Tools/TestWebKitAPI/glib/WebKitGLib/WebExtensionUtilities.cpp: Copied from Source/WebKit/UIProcess/API/glib/WebKitPrivate.h.
(TestWebKitAPI::Util::makePNGData):
* Tools/TestWebKitAPI/glib/WebKitGLib/WebExtensionUtilities.h: Copied from Source/WebKit/UIProcess/API/glib/WebKitPrivate.h.

Canonical link: https://commits.webkit.org/301078@main

mcatanzaro commented on Source/WebCore/platform/graphics/gtk/IconGtk.cpp:
```diff
@@ -35,7 +35,7 @@
 
-Icon::Icon(GIcon* icon)
+Icon::Icon(GRefPtr<GIcon>&& icon)
     : m_icon(icon)
```
```suggestion
    : m_icon(WTFMove(icon))
```

mcatanzaro commented on Source/WebKit/Shared/API/APIData.h:
```diff
@@ -76,6 +82,14 @@ class Data : public ObjectImpl<API::Object::Type::Data> {
 #endif
 
+#if PLATFORM(GTK)
+    static Ref<Data> createWithoutCopying(GRefPtr<GBytes>&& bytes)
+    {
+        auto span = WTF::span(bytes);
+        return createWithoutCopying(span, [bytes = WTFMove(bytes)] () { });
+    }
```
OK, I see this works because the destroy function receives ownership of the GRefPtr. This was smart. (Maybe too smart: I might have added a short comment to explain it!)

mcatanzaro commented on Source/WebKit/Shared/Extensions/gtk/WebExtensionUtilitiesGtk.cpp:
```diff
@@ -0,0 +1,60 @@
+    auto* monitors = gdk_display_get_monitors(display);
+    unsigned currentMonitor = 0;
+    while (auto* item = g_list_model_get_item(monitors, currentMonitor++)) {
+        auto* monitor = GDK_MONITOR(item);
+        ASSERT(GDK_IS_MONITOR(monitor));
```
`GDK_MONITOR()` will itself assert, since WebKit does not disable cast checks.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitError.h.in:
```diff
@@ -158,6 +161,36 @@ typedef enum {
+ * 
+ * Since: 2.52
+ */
+typedef enum {
+    WEBKIT_WEB_EXTENSION_ERROR_UNKNOWN = 899,
+    WEBKIT_WEB_EXTENSION_ERROR_RESOURCE_NOT_FOUND = 800,
+    WEBKIT_WEB_EXTENSION_ERROR_INVALID_RESOURCE_CODE_SIGNATURE = 801,
+    WEBKIT_WEB_EXTENSION_ERROR_INVALID_MANIFEST = 802,
+    WEBKIT_WEB_EXTENSION_ERROR_UNSUPPORTED_MANIFEST_VERSION = 803,
+    WEBKIT_WEB_EXTENSION_ERROR_INVALID_MANIFEST_ENTRY = 804,
+    WEBKIT_WEB_EXTENSION_ERROR_INVALID_DECLARATIVE_NET_REQUEST_ENTRY = 805,
+    WEBKIT_WEB_EXTENSION_ERROR_INVALID_BACKGROUND_PERSISTENCE = 806,
+    WEBKIT_WEB_EXTENSION_ERROR_INVALID_ARCHIVE = 807,
+} WebKitWebExtensionError;
```
Where do these integer values come from? Would be good to add a hint, so future developers don't mess up when adding a new error.

ItsJamie9494 replied:
Arbitrarily chosen to try and avoid a conflict with another error domain. The internal errors that WebExtension uses are simply just integers like 1, 2, 3...which probably shouldn't be exported publicly.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.h.in:
```diff
@@ -0,0 +1,124 @@
+webkit_web_extension_new                                     (GFile   *extension_path,
+                                                              GError **error);
+
+WEBKIT_API GFile *
+webkit_web_extension_get_path                                (WebKitWebExtension* extension);
```
```suggestion
WEBKIT_API GFile *
webkit_web_extension_get_path                                (WebKitWebExtension *extension);
```

Hm, I would have expected a `get_path()` function to return a `const char *` or `char *`. What do you think about calling this `get_file()` instead?

Actually, using GFile is probably a little overkill. It's more common to just use paths in APIs, right? So even more interesting: what do you think about changing this, and also `webkit_web_extension_new()` and `WebKitWebExtension:path`, to just use a path instead of a GFile?

ItsJamie9494 replied:
For public API, we can certainly use a path. My main thought was that most cases of using this will probably have a GFile already, but looking at how Ephy works, I just end up creating a GFile to pass into `_new()`, so changing this makes a lot of sense

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
```diff
@@ -0,0 +1,1298 @@
+
+/**
+ * webkit_web_extension_new:
+ * @extension_path: A #GFile pointing to the folder containing the extension manifest and resources
+ * @error: return location for error or %NULL to ignore
```
So just to make sure you're aware, this is gtk-doc syntax. We should eventually convert it to gi-docgen syntax. But gtk-doc syntax is still widely-used throughout our API docs, and we use gi-docgen syntax almost nowhere, so you don't actually have to worry about this yourself.

ItsJamie9494 replied:
Yes, I am aware. I occasionally try and remember to use gi-docgen syntax, but since gtk-doc is so prominent throughout this codebase I sometimes forget.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
```diff
@@ -0,0 +1,1298 @@
+ * Get the parsed manifest version, or `0` if there is no
+ * version specified in the manifest.
+ *
+ * An ``WKWebExtensionErrorUnsupportedManifestVersion`` error will be
```
```suggestion
 * A ``WKWebExtensionErrorUnsupportedManifestVersion`` error will be
```

mcatanzaro reviewed (commented):
OK, I finally got around to reviewing this. Better late than never! I like this. Thanks.

I left a few minor comments. Two things seem actually important. First, are you planning to change:

```
WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_WK_WEB_EXTENSIONS PRIVATE ${ENABLE_EXPERIMENTAL_FEATURES})
```

to:

```
WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_WK_WEB_EXTENSIONS PRIVATE ON)
```

? Otherwise, the new API is not actually usable. Unless you have some really good reason that it needs to be disabled at build time, we should enable it.

Second, I left another comment regarding whether we should use GFile in the public API, or just change it to a C string.
